cni/pkg/nodeagent/cni-watcher.go

```diff
@@ -0,0 +1,181 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodeagent
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/netip"
+
+	"github.com/containernetworking/cni/pkg/skel"
+
+	pconstants "istio.io/istio/cni/pkg/constants"
+	"istio.io/istio/cni/pkg/pluginlistener"
+	"istio.io/istio/pkg/network"
+)
+
+// Just a composite of the CNI plugin add event struct + some extracted "args"
+type CNIPluginAddEvent struct {
+	CmdAddEvent  skel.CmdArgs
+	PodName      string
+	PodNamespace string
+	IPs          []IPConfig
+}
+
+// IPConfig contains an interface/gateway/address combo defined for a newly-started pod by CNI.
+// This is "from the horse's mouth" so to speak and will be populated before Kube is informed of the
+// pod IP.
+type IPConfig struct {
+	Interface *int
+	Address   net.IPNet
+	Gateway   net.IP
+}
+
+type CniPluginServer struct {
+	cniListenServer       *http.Server
+	cniListenServerCancel context.CancelFunc
+	handlers              K8sHandlers
+	dataplane             MeshDataplane
+
+	sockAddress string
+	ctx         context.Context
+}
+
+func startCniPluginServer(ctx context.Context, pluginSocket string,
+	handlers K8sHandlers,
+	dataplane MeshDataplane,
+) *CniPluginServer {
+	ctx, cancel := context.WithCancel(ctx)
+	mux := http.NewServeMux()
+	s := &CniPluginServer{
+		handlers:  handlers,
+		dataplane: dataplane,
+		cniListenServer: &http.Server{
+			Handler: mux,
+		},
+		cniListenServerCancel: cancel,
+		sockAddress:           pluginSocket,
+		ctx:                   ctx,
+	}
+
+	mux.HandleFunc(pconstants.CNIAddEventPath, s.handleAddEvent)
+	return s
+}
+
+func (s *CniPluginServer) Stop() {
+	s.cniListenServerCancel()
+}
+
+// startCNIPluginListener starts up a UDS server which receives events from the CNI chain plugin.
+func (s *CniPluginServer) Start() error {
+	if s.sockAddress == "" {
+		return fmt.Errorf("no socket address provided")
+	}
+	log.Info("Start a listen server for CNI plugin events")
+	unixListener, err := pluginlistener.NewListener(s.sockAddress)
+	if err != nil {
+		return fmt.Errorf("failed to create CNI listener: %v", err)
+	}
+	go func() {
+		if err := s.cniListenServer.Serve(unixListener); network.IsUnexpectedListenerError(err) {
+			log.Errorf("Error running CNI listener server: %v", err)
+		}
+	}()
+
+	go func() {
+		<-s.ctx.Done()
+		if err := s.cniListenServer.Close(); err != nil {
+			log.Errorf("CNI listen server terminated with error: %v", err)
+		} else {
+			log.Debug("CNI listen server terminated")
+		}
+	}()
+
+	return nil
+}
+
+func (s *CniPluginServer) handleAddEvent(w http.ResponseWriter, req *http.Request) {
+	if req.Body == nil {
+		log.Error("empty request body")
+		http.Error(w, "empty request body", http.StatusInternalServerError)
```

Comment:
No early return for empty response body?

Reply:
Also, should this better to be BadRequest?

Reply:
yep! ty.